### PR TITLE
Expose normalized transport-error detail (ErrorString/ErrorSymbol/ErrorCode)

### PR DESCRIPTION
**Pull request by matthargett**
Closes the transport-error information gap on the Apple and libcurl backends: failures like DNS errors, refused connections, TLS problems, and missing local files previously completed with `StatusCode() == 0` and **no other information**. The Apple backend discarded the `NSError` (the existing `// TODO: Consider logging or otherwise exposing the error message...`), and the curl backend caught and dropped its own failure exception (necessarily, to protect its `std::thread`), so consumers like JsRuntimeHost's `fetch`/`XMLHttpRequest` polyfills can only say "network request failed". For applications that embed Babylon Native and live off field crash reports, every failure class is currently indistinguishable.

### API (purely additive)

```cpp
std::string_view ErrorString();  // "<domain>:<symbol>(<code>): <detail>", empty if no transport failure
std::string_view ErrorSymbol();  // stable token, e.g. "CURLE_COULDNT_RESOLVE_HOST", "NSURLErrorTimedOut"
int32_t          ErrorCode();    // raw numeric platform code (CURLcode / NSError code)
```

`SendAsync()` still completes normally with status 0 on transport failure, so **existing consumers are completely unaffected** until they read the new accessors (same opt-in shape as `StatusText()` in #30).

### Normalized, greppable shape

`<domain>` and `<symbol>` are stable ASCII tokens, so observability pipelines (Splunk etc.) can filter classes of failure with exact substrings like `"curl:CURLE_COULDNT_RESOLVE_HOST"` or `"nsurl:NSURLErrorTimedOut"`; `<detail>` carries the platform's human-readable message including host/port/path specifics. Real output from this branch (macOS):

```
nsurl:NSURLErrorCannotConnectToHost(-1004): Could not connect to the server.
nsurl:NSURLErrorCannotFindHost(-1003): A server with the specified hostname could not be found.
nsurl:NSURLErrorFileDoesNotExist(-1100): The requested URL was not found on this server.
nsurl:NSURLErrorServerCertificateUntrusted(-1202): The certificate for this server is invalid. You might be connecting to a server that is pretending to be "expired.badssl.com" ...
urllib:AppResourceNotFound(0): no bundled resource for 'app:///missing.js'
```

and on Linux/curl, `CURLOPT_ERRORBUFFER` detail is preferred over generic `curl_easy_strerror` text:

```
curl:CURLE_COULDNT_CONNECT(7): Failed to connect to 127.0.0.1 port 47651 after 0 ms: Couldn't connect to server
curl:CURLE_FILE_COULDNT_READ_FILE(37): Couldn't open file /tmp/missing.bin
```

Notes:
- Apple: `NSURLErrorDomain` codes map to stable symbols; other domains pass through verbatim with a synthesized symbol. The underlying-`NSError` chain is walked (bounded, deduped by code) so POSIX-level specifics surface when present. The `<detail>` text is OS-localized on Apple; domain/symbol/code are the stable filterable parts.
- The missing-`app:///`-resource case (probably the most common asset-loading failure in shipped apps) is now distinguishable from a network failure via `urllib:AppResourceNotFound(0)`.
- The invalid-URL `Open()` throw on Apple now includes the offending URL string.

### Scope

Apple (`NSURLSession`) and Linux (`libcurl`) backends per the priority consumers; Windows and Android backends keep reporting empty/zero through the shared base (the `SetError` plumbing in `UrlRequest_Base.h` is ready for them as a follow-up).

### Tests

`Tests/UrlRequestErrorReporting.cpp` adds 7 offline-deterministic gtest cases: local-file success (no error), missing local file, connection refused (ephemeral bound-then-closed loopback port), NXDOMAIN (RFC 6761 `.invalid` TLD), grammar-shape regex, request-reuse clears prior error, and missing `app:///` resource (Apple). Tests are built only when UrlLib is the top-level project (`URLLIB_TESTS`, default ON for desktop top-level builds, OFF when consumed via FetchContent), so JsRuntimeHost et al. see no change. I looked at how bun, nodejs, and WPT tests around this area.

CI for this repo is added separately in #32 (Win32 / macOS / Linux, mirroring the JsRuntimeHost workflow conventions). **Landing order: this PR first, then #32** — its `Run Tests` steps execute the `UrlLibTests` target introduced here.

**Verified:** built and ran the suite on macOS arm64 (7/7 passing); curl TU compiles clean with `-Wall -Wextra` (clang); Linux (gcc + clang), macOS, and Win32 execution via the #32 workflows running against this branch on the fork mirror (rebeckerspecialties/UrlLib#1): https://github.com/rebeckerspecialties/UrlLib/actions/runs/27312871316

### Context / follow-ups

This is step 1 of improving fetch-failure telemetry / crash reporting for BabylonNative/JsRuntimeHost (see the fetch polyfill added in BabylonJS/JsRuntimeHost#188, whose transport-failure rejection is currently the constant string `"fetch: network request failed"`). To have actionable support information for triaging/fixing, these errors need to have the non-PII fidelity *and* reach the top-level abort/reject handler that Sentry/bugsnag/etc hook.

Follow-up in JsRuntimeHost as the next hop in the exception/reject accuracy chain: reject with a `TypeError` carrying a stable message plus `cause`/`code` properties populated from these accessors, mirroring how #30's `StatusText()` was consumed by BabylonJS/JsRuntimeHost#195. 

Noticed while here (not addressed in this PR): `UrlRequest::Abort()` only has an effect on the Windows backend today, and the curl backend rejects `POST` with "Not implemented".

### CODE REVIEW

File: Tests/UrlRequestErrorReporting.cpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        auto done = std::make_shared<std::promise<void>>();
+        auto future = done->get_future();
+
+        request.SendAsync().then(arcana::inline_scheduler, arcana::cancellation::none(),
+            [done](const arcana::expected<void, std::exception_ptr>&) {
+                done->set_value();
+            });
+
+        return future.wait_for(std::chrono::seconds{120}) == std::future_status::ready;
+    }
```
[the remark this answers is not in the document]
Comment (matthargett) [reply Tests/UrlRequestErrorReporting.cpp]: Adopted in 06b5287: the wait is now 30s and the helper calls `request.Abort()` before failing on timeout.

One honest caveat from tracing the backends while making this change: `UrlRequest::Abort()` only cancels `m_cancellationSource`, which the curl and NSURLSession backends never observe — so cancellation is only actually effective on the Windows backend today, and on the platforms this CI runs the destructor can still block until the transport gives up on its own. The real bound for this suite is that every scenario is offline-deterministic (loopback refusal, RFC 6761 `.invalid` NXDOMAIN, local files), so completion is fast by construction; the 30s wait and ctest `--timeout` are backstops. Making `Abort()` interrupt the Apple/curl transports is part of the follow-up chain this PR feeds (fetch `AbortSignal` wiring in JsRuntimeHost).

File: Tests/UrlRequestErrorReporting.cpp [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        address.sin_port = 0;
+        EXPECT_EQ(::bind(fd, reinterpret_cast<const sockaddr*>(&address), sizeof(address)), 0);
+
+        socklen_t addressLength = sizeof(address);
+        EXPECT_EQ(::getsockname(fd, reinterpret_cast<sockaddr*>(&address), &addressLength), 0);
+        const uint16_t port = ntohs(address.sin_port);
+
+        ::close(fd);
+        return port;
+    }
```
[the remark this answers is not in the document]
Comment (matthargett) [reply Tests/UrlRequestErrorReporting.cpp]: Both points adopted in 06b5287, with one twist based on my on-device testing. The helper is now a move-only RAII `RefusingPort` whose acquisition returns `std::optional` with `ASSERT_TRUE` at call sites, so setup failure stops the test instead of continuing on a bad fd.

On the hold-the-bind suggestion: it turns out to be platform-dependent. On Linux a SYN to a bound-but-not-listening port is refused with RST, so there the socket is kept bound for the test's lifetime exactly as you describe — race-free. On Darwin, though, that same SYN is silently **dropped**: the connect attempt times out (`NSURLErrorTimedOut(-1001)`) instead of being refused (`-1004`) — verified by running it. So on Apple the port number is reserved via bind/getsockname and then closed before connecting, accepting the microseconds-wide reuse window. The class documents the split.

File: Tests/UrlRequestErrorReporting.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+    std::filesystem::path WriteTempFile(const char* name, const char* contents)
+    {
+        const auto path = std::filesystem::temp_directory_path() / name;
+        std::ofstream stream{path, std::ios::trunc};
+        stream << contents;
+        return path;
+    }
```
[the remark this answers is not in the document]
Comment (matthargett) [reply Tests/UrlRequestErrorReporting.cpp]: Adopted in 06b5287 as an RAII `TempFile` that suffixes the process id (so parallel jobs sharing a temp directory get distinct paths) and removes the file in its destructor; the missing-file test reuses the same fixture with null contents for a guaranteed-absent path. One note on the suggested API: `std::filesystem::unique_path` is Boost-only — it never made it into `std::filesystem` — hence the pid-suffixed naming.

**Comment by matthargett** [comment]
CI evidence while first-time-workflow approval is pending here: this branch is mirrored as a draft PR on our fork (rebeckerspecialties/UrlLib#1), where the workflows ran green across all three jobs — `Ubuntu_gcc`, `Ubuntu_clang`, and `macOS_Xcode164`: https://github.com/rebeckerspecialties/UrlLib/actions/runs/27312401213

That includes the first real execution of the libcurl error path (previously only syntax-checked): the Linux jobs pass the exact-symbol assertions — `CURLE_COULDNT_CONNECT(7)` with the host detail from `CURLOPT_ERRORBUFFER`, `CURLE_COULDNT_RESOLVE_HOST(6)`, and `CURLE_FILE_COULDNT_READ_FILE(37)`.

Also note the CI was restructured in 4ccfaba to match the JsRuntimeHost/BabylonNative conventions: a `ci.yml` orchestrator calling reusable `build-linux.yml` / `build-macos.yml` workflows (`checkout@v5`, Ninja with gcc+clang jobs, pinned Xcode via `xcode-select`, test binaries run directly), so extending to Win32/UWP/Android jobs later follows the same shape as the sibling repos.

**Comment by matthargett** [comment]
Win32 is now in the CI matrix per Microsoft-project table stakes: `build-win32.yml` mirrors JsRuntimeHost's (windows-2022 / VS2022 generator pin, `-A` platform input, RelWithDebInfo with `/m`, crash-dump staging + artifact upload on failure), with a `Win32_x64` job in `ci.yml`.

All four jobs green on the fork mirror: https://github.com/rebeckerspecialties/UrlLib/actions/runs/27312871316

On Windows the suite builds and runs (Winsock fixture variant, `_getpid`, `file:///C:/...` URL formation): the local-file success case passes for real — validating the Win32 backend's file path and the empty-`ErrorString()` no-error contract — while the five assertions that need transport-error detail `GTEST_SKIP` with an explicit "the Windows backend does not populate transport-error detail yet" message, so the platform gap stays visible in test output without a permanently red job. Populating the Windows backend (the `winrt::hresult_error` catch sites in `UrlRequest_Windows_Shared.h`) flips those skips to live coverage as the follow-up.

**Comment by matthargett** [comment]
Repo-infrastructure split: the CI workflows have been moved out of this PR into #32 so they can be reviewed on their own merits — this PR is now scoped to the error-reporting code, tests, and docs (the `.github/` files are gone from the diff as of c2c7fce).

**Landing order: this PR first, then #32.** #32's `Run Tests` steps execute the `UrlLibTests` target introduced here, so it depends on this one (noted in its description).

The fork mirror (rebeckerspecialties/UrlLib#1) now validates the combined state of both PRs — it targets the #32 branch as its base with this branch as the head, so every push to either PR re-runs the full Win32/macOS/Linux matrix against exactly what `main` will look like after both land.

**Comment by matthargett** [comment]
Correction to the mirror note above: the validation mirror is now rebeckerspecialties/UrlLib#2 (a fork-only branch combining this PR with #32's workflows — base-side-only workflow files turn out not to trigger `pull_request` runs, so the combined-head shape it is). Post-split state validated green across all four jobs: https://github.com/rebeckerspecialties/UrlLib/actions/runs/27313677186

**Comment by bkaradzic-microsoft** [comment]
Landed on `main` in e86ffb3 via #33 (squash), together with maintainer review fixes applied on top of your work — your authorship is credited via `Co-authored-by`.

**Why a separate PR rather than merging this one directly:** this PR's head fork is org-owned, so GitHub blocks maintainer-edit pushes onto its branch (despite `maintainerCanModify == true`), and the repo is squash-only. #33 carried your commits plus the review fixes below.

**Review fixes applied:**
- Documented that a transport failure is signaled by a non-empty `ErrorString()`/`ErrorSymbol()`, **not** `ErrorCode() != 0` — since `AppResourceNotFound` legitimately reports code `0` (header + README), with a test asserting the code-0 contract.
- Apple: added `ToStdString()` so a nil `NSString` (`localizedDescription`/`domain`) can't construct `std::string{nullptr}` (UB) when building the error detail / walking the underlying-error chain.
- Unix: renamed the synthesized getinfo token `CURLE_GETINFO_FAILED` → `GetInfoFailed` so it doesn't masquerade as a real `CURLcode`.
- Tests: loosened the curl `.invalid` DNS assertion to require a `curl:` transport error with a non-zero code instead of pinning `CURLE_COULDNT_RESOLVE_HOST(6)` — tolerates proxy / DNS-hijacking resolvers; exact symbols stay covered by the connection-refused and missing-file cases.

Validated by building/running `UrlLibTests` on Win32 (success case passes; transport-detail cases `GTEST_SKIP` as before).

Thanks @matthargett! Closing as landed via #33. CI workflows (#32) are the next hop.
